docker.yml

```diff
@@ -0,0 +1,11 @@
+.docker_build_publish:
+  stage: docker_build
+  tags:
+    - meao
+    - aws
+  script:
+    - cd $subdir
+    - docker build -t $image_name:latest -t $image_name:$version --build-arg version=$version $build_args .
```

Comment:
do we always want `latest`?
any thoughts @pmac @glogiotatidis?

Reply:
we've done this differently on different projects. Sometimes `:latest` meant most recent production image, and others meant most recent image from master and used a separate `:latest-prod` for the production one (e.g. bedrock). So I guess it would be nice if we could standardize that. But it is definitely true that not every docker build in CI should get the `:latest` tag.

Reply:
It seems possible to do some configurable tagging. Maybe something like:

- always get a version
- by default get 'latest' (overridable)
- be able to add 'latest-prod'
- be able to get 'all-versions' (which would the the split on . thing, giving you, 1.2.3, 1.2 and 1)
- be able to add any other arbitrary static label (basically reusing latest-prod setup)

Does that sound good enough for now?

Reply:
what it this thing always tagged with a specific tag, then called a script that is expected to exist in the repo to do whatever tagging it wants, and returns the tags it wants pushed. e.g.:

```
- docker build -t $image_name:pipeline-latest ...
- image_tags=$(./bin/tag-image $image_name:pipeline-latest)
- for img_tag in $image_tags; do docker push $img_tag; done
```

Reply:
Also, where does `$version` come from? Variables defined in the pipeline? It can be the git hash right? It can probably just be the `$CI_COMMIT_SHORT_SHA` value by default.

Reply:
In this case it's whatever version wrangler has picked for their release.  I'm trying to just copy their style, rather than make my own.

TBH, I'm not the biggest fan of versioning via the short sha/commit ids in general.  It's hard to figure out ordering by just looking at containers/pipeline runs.  I know a ton of our repos do it, and there's probably not a reason to change right away, so I'll try to make this pipeline support that pattern, but don't think it should be the default.

Reply:
What tagging with the git hash enables is easily finding the image you want for the version of the code you want. We don't version site releases because semantic version is meaningless for the sites. We do usually also tag prod with a date tag (e.g. `2020-04-16.2`), and bedrock includes the branch since the build for master branch contains different data than the prod branch. We have different needs. If this is not for use by our non-wrangler projects then great. If it is then it should be able to do what we need.

Reply:
@pmac you're right. I'm gonna ask for a mulligan on this one I think, and file an issue for making this good. I'll make it a little bit more flexible for now (probably just make the latest bit an env var and call it good). We should design it better than what I have done so far.

#11

Reply:
I'm going to merge this work as is, I think as of right now it's blocking @glogiotatidis work. Tried to make this a bit more explicit about the shortcomings, which I think ought to address the concerns here.